### PR TITLE
Fix PostgreSQL compound key discovery

### LINKED ISSUE #56420: QGIS 3.34 is not able to figure out PostgreSQL relations when there are build using compound keys
### What is the bug or the crash?

QGIS 3.34 is not able to figure out PostgreSQL relations when there are build using compound keys.

When I click the button "Discover relations" under "Project" > "Properties" > "Relations" it shows:

Screenshot from 2024-02-20 09-41-47

Where you can see that all children relations start with "bush_", where there should be 2 of each, or better one of each with each time the two field used as the compound key.    

Also, there are two "replaced_" entries vs only one "replacing_" where there should also be two and their respective structure should be mirror 2-by-2 but they are not.

My wish is to have:
| Name | Referencing layer | Referencing field | Referenced layer | Referenced field | Strength |
|---|---|---|---|---|---|
| bush_entity_type_id_fkey | bush | entity_type, entity_id | parent | entity_type, id | Association |
| pond_entity_type_id_fkey | pond | entity_type, entity_id | parent | entity_type, id | Association |
| rock_entity_type_id_fkey | rock | entity_type, entity_id | parent | entity_type, id | Association |
| replaced_entity_type_id_fkey | replacement | replaced_entity_type, replaced_entity_id | parent | entity_type, entity_id | Association |
| replacing_entity_type_id_fkey | replacement | replacing_entity_type, replacing_entity_id | parent | entity_type, entity_id | Association |

### Steps to reproduce the issue

PostGIS 16+3.4 database sample:

```sql
CREATE TABLE public.entity
(
    id INT NOT NULL GENERATED ALWAYS AS IDENTITY,
    entity_type TEXT NOT NULL,
    created_at TIMESTAMPTZ DEFAULT NOW(),
    PRIMARY KEY (entity_type, id),
    CONSTRAINT entity_type_chk
        CHECK (entity_type IN ('bush', 'pond', 'rock'))
);

CREATE TABLE public.bush
(
    id INT PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
    entity_type TEXT NOT NULL DEFAULT 'bush',
    entity_id INT NOT NULL,
    species TEXT,
    diameter INT,
    geom geometry(Point,4326),
    CONSTRAINT bush_entity_type_id_fkey FOREIGN KEY (entity_type, entity_id)
        REFERENCES public.entity (entity_type, id) MATCH SIMPLE
        ON UPDATE NO ACTION
        ON DELETE NO ACTION
);

CREATE TABLE public.pond
(
    id INT PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
    entity_type TEXT NOT NULL DEFAULT 'pond',
    entity_id INT NOT NULL,
    common_name TEXT,
    surface INT,
    geom geometry(MultiPolygon,4326),
    CONSTRAINT bush_entity_type_id_fkey FOREIGN KEY (entity_type, entity_id)
        REFERENCES public.entity (entity_type, id) MATCH SIMPLE
        ON UPDATE NO ACTION
        ON DELETE NO ACTION
);

CREATE TABLE public.rock
(
    id INT PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
    entity_type TEXT NOT NULL DEFAULT 'rock',
    entity_id INT NOT NULL,
    nickname TEXT,
    mineral TEXT,
    geom geometry(MultiPolygon,4326),
    CONSTRAINT bush_entity_type_id_fkey FOREIGN KEY (entity_type, entity_id)
        REFERENCES public.entity (entity_type, id) MATCH SIMPLE
        ON UPDATE NO ACTION
        ON DELETE NO ACTION
);

CREATE TABLE public.replacement
(
    id integer PRIMARY KEY GENERATED BY DEFAULT AS IDENTITY,
    replaced_entity_type TEXT,
    replaced_entity_id INT,
    replacing_entity_type TEXT,
    replacing_entity_id INT,
    CONSTRAINT replaced_entity_type_id_fkey FOREIGN KEY (replaced_entity_type, replaced_entity_id)
        REFERENCES public.entity (entity_type, id) MATCH SIMPLE
        ON UPDATE NO ACTION
        ON DELETE NO ACTION,
    CONSTRAINT replacing_entity_type_id_fkey FOREIGN KEY (replacing_entity_type, replacing_entity_id)
        REFERENCES public.entity (entity_type, id) MATCH SIMPLE
        ON UPDATE NO ACTION
        ON DELETE NO ACTION
);


INSERT INTO public.entity(entity_type) VALUES
  ('bush'),
  ('bush'),
  ('bush'),
  ('pond'),
  ('pond'),
  ('pond'),
  ('rock'),
  ('rock'),
  ('rock');

INSERT INTO public.bush(species, entity_id, diameter, geom) VALUES
  ('Quercus Ilex', 1, 60, ST_GeomFromText('POINT(9 7)', 4326)),
  ('Fagus Sylvatica', 2, 55, ST_GeomFromText('POINT(15 1)', 4326)),
  ('Acer Pseudoplatanus', 3, 70, ST_GeomFromText('POINT(1 6)', 4326));

INSERT INTO public.pond(common_name, entity_id, surface, geom) VALUES
  ('Blue pond', 4, 60, ST_GeomFromText('POLYGON((0 4, 0 3, 3 3, 3 4, 0 4))', 4326)),
  ('Royal beauty', 5, 55, ST_GeomFromText('POLYGON((2 9, 6 9, 7 11, 3 11, 2 9))', 4326)),
  ('Prinz Hertzpold', 6, 70, ST_GeomFromText('POLYGON((14 6, 20 6, 17 8, 14 6))', 4326));

INSERT INTO public.rock(nickname, entity_id, mineral, geom) VALUES
  ('Rocky Balboa', 7, 'basalt', ST_GeomFromText('POLYGON((14 12, 15 12, 15 13, 14 13, 14 12))', 4326)),
  ('Hard rock', 8, 'limestone', ST_GeomFromText('POLYGON((7 8, 8 8, 8 9, 7 9, 7 8))', 4326)),
  ('The Rock', 9, 'granite', ST_GeomFromText('POLYGON((10 2, 11 2, 11 3, 10 3, 10 2))', 4326));

```

Set up the database connection in QGIS, add the 5 layers (two are non-geometric layers).

Click the button "Discover relations" under "Project" > "Properties" > "Relations" it should display the same screenshot as shown above.

My ultimate goal is to be able to feed the `replacement` table from QGIS by linking (many-to-many) replaced-replacing features, each time from the three child layers. See [here](https://gis.stackexchange.com/questions/476254/managing-postgis-many-to-many-relationships-between-several-tables-with-differen) for a detail explanation.

### Versions

QGIS version | 3.34.3-Prizren | QGIS code revision | 47373234acd
-- | -- | -- | --
Qt version | 5.15.3
Python version | 3.10.12
GDAL/OGR version | 3.4.1
PROJ version | 8.2.1
EPSG Registry database version | v10.041 (2021-12-03)
GEOS version | 3.10.2-CAPI-1.16.0
SQLite version | 3.37.2
PDAL version | 2.3.0
PostgreSQL client version | 14.10 (Ubuntu 14.10-0ubuntu0.22.04.1)
SpatiaLite version | 5.0.1
QWT version | 6.1.4
QScintilla2 version | 2.11.6
OS version | Ubuntu 22.04.4 LTS
  |   |   |  
Active Python plugins
plugin_reloader | 0.9.3
swissgeodownloader | 2.0.1
kart | 1.0.14
Serval | 3.32.0
qgis-maptiler-plugin | 3.2
interlis | 0.14.8
QuickWKT | 3.1
changeDataSource | 3.1
linking_relation_editor | v1.2.0
QgisModelBaker | v7.8.1
helloworld | 0.1
profiletool | 4.2.6
qfieldsync | v4.7.1
QuickOSM | 2.2.3
pluginbuilder3 | 3.2.1
quick_map_services | 0.19.34
gmsh | version 0.2.2
oneatlas-qgis-plugin-1_0_1 | 1.3.0
crayfish | 3.6.0
db_manager | 0.1.20
processing | 2.12.99
MetaSearch | 0.3.6
grassprovider | 2.12.99

</body></html>

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by strk**
Closes GH-56420

### CODE REVIEW

**Comment by strk** [comment]
Tests for relation discovery seem to exist but are based on referenced keys being first attributes of the referenced table. I'm going to update that one.

To run:

> QGIS_PREFIX_PATH=${builddir}/output PYTHONPATH=${builddir}/output/python:$PYTHONPATH python tests/src/python/test_provider_postgres.py TestPyQgsPostgresProvider.testValidLayerDiscoverRelationsComposite

File: src/providers/postgres/qgspostgresprovider.cpp [hunk 1]
```diff
@@ -5260,13 +5259,11 @@ QList<QgsRelation> QgsPostgresProvider::discoverRelations( const QgsVectorLayer
       refSchema = refSchema.mid( 1, refSchema.length() - 2 );
     }
     const QString refColumn = sqlResult.PQgetvalue( row, 4 );
-    const QString position = sqlResult.PQgetvalue( row, 5 );
-    const QList<QgsVectorLayer *> foundLayers = searchLayers( layers, mUri.connectionInfo( false ), refSchema, refTable );
-    if ( ( position == QLatin1String( "1" ) ) || ( !refTableFound.contains( refTable ) ) )
+    // try to find if we have layers for the referenced table
+    const auto foundLayers = searchLayers( layers, mUri.connectionInfo( false ), refSchema, refTable );
```
Comment (nyalldawson) [inline src/providers/postgres/qgspostgresprovider.cpp:5263]: this makes things less readable, let's not change it:

```suggestion
    const QList<QgsVectorLayer *> foundLayers = searchLayers( layers, mUri.connectionInfo( false ), refSchema, refTable );
```

**Comment by strk** [comment]
Bug confirmed by CI: https://github.com/qgis/QGIS/actions/runs/9446127636/job/26016503531?pr=57653#step:13:2257
Now I'll force-push a version of the branch with the fix in place

File: src/providers/postgres/qgspostgresprovider.cpp [hunk 2]
```diff
@@ -5265,13 +5264,11 @@ QList<QgsRelation> QgsPostgresProvider::discoverRelations( const QgsVectorLayer
...
```
Comment (nyalldawson) [inline src/providers/postgres/qgspostgresprovider.cpp:5268]:
```suggestion
    const QList<QgsVectorLayer *> foundLayers = searchLayers( layers, mUri.connectionInfo( false ), refSchema, refTable );
```

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**